Add climate fan modes for Airzone Cloud Aidoo devices

Noltari opened the pull request:
## Proposed change
Add climate fan modes for Airzone Cloud Aidoo devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Noltari committed:
airzone_cloud: climate: add Aidoo speeds support

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

joostlek commented on homeassistant/components/airzone_cloud/climate.py:
```diff
@@ -262,6 +285,9 @@ async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
 class AirzoneAidooClimate(AirzoneAidooEntity, AirzoneDeviceClimate):
     """Define an Airzone Cloud Aidoo climate."""
 
+    _speeds: dict[int, str] = {}
+    _speeds_reverse: dict[str, int] = {}
```
Please set these in the constructor, having mutable dicts outside of the constructor can have unwanted side effects

Noltari replied:
In fact we don't need to init them here, since `_set_fan_speeds` is already doing that, so I will remove the `= {}` from here.
Sorry, I just copied this from the airzone integration: https://github.com/home-assistant/core/blob/c83388fd2da3fab062dec3fc3c3ebc8013ca478b/homeassistant/components/airzone/climate.py#L118-L119

Noltari replied:
> In fact we don't need to init them here, since `_set_fan_speeds` is already doing that, so I will remove the `= {}` from here. Sorry, I just copied this from the airzone integration:
> 
> https://github.com/home-assistant/core/blob/c83388fd2da3fab062dec3fc3c3ebc8013ca478b/homeassistant/components/airzone/climate.py#L118-L119

Done in cdbe317185844534075504d610aec0a92beb640e

joostlek commented on homeassistant/components/airzone_cloud/climate.py:
```diff
@@ -278,9 +304,58 @@ def __init__(
+        azd_speeds: dict[int, int] = self.get_airzone_value(AZD_SPEEDS)
+        max_speed = max(azd_speeds)
+
+        speeds: dict[int, str]
+        if _speeds := FAN_SPEED_MAPS.get(max_speed):
+            speeds = _speeds
+        else:
+            speeds = {}
```
I don't know if its possible, but can we give the names some more context? We now have `speeds`, `_speeds`, `self._speeds` and its kinda confusing

Noltari replied:
Done in cdbe317185844534075504d610aec0a92beb640e

joostlek commented on homeassistant/components/airzone_cloud/climate.py:
```diff
@@ -278,9 +304,58 @@ def __init__(
 
         self._async_update_attrs()
 
+    def _set_fan_speeds(self) -> None:
```
I have the feeling this function is probably a tad too complex, but I have no better way.

But if I understand correctly, we will create fan modes depending on what the fan can do. Is this maybe something to move to the lib? Is this HA specific?

Noltari replied:
This is pretty similar to https://github.com/home-assistant/core/blob/c83388fd2da3fab062dec3fc3c3ebc8013ca478b/homeassistant/components/airzone/climate.py#L152-L171 but as opposed to the local integration, in which all devices seems to support 0 (auto) speed, some older cloud devices can only be set to manual speeds...

Noltari replied:
Also, for me this is more Home Assistant specific, since this function allows us to expose common fan modes such as `LOW`, `MEDIUM` and `HIGH`, but it also adds the rest of the speeds as a percentage. This is done so these speeds can be exposed to Alexa, Homekit...
See https://github.com/home-assistant/core/pull/92840#issuecomment-1542049704

Noltari committed:
Add suggested changes

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into airzone-cloud-climate-speeds

jesusvila commented:
I'm interested in this function. How can I test it?

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into airzone-cloud-climate-speeds

Noltari committed:
airzone_cloud: climate: rename _set_fan_speeds()

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
airzone_cloud: climate: drop BASE_FAN_SPEEDS

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari commented on homeassistant/components/airzone_cloud/climate.py:
```diff
@@ -54,6 +61,22 @@
+    | {
+        2: FAN_MEDIUM,
+        3: FAN_HIGH,
+    },
```
Done in 1feff044ffb5fa9a53c77b06e531587d3216da4d

Noltari commented on homeassistant/components/airzone_cloud/climate.py:
```diff
@@ -54,6 +61,22 @@
+BASE_FAN_SPEEDS: Final[dict[int, str]] = {
+    0: FAN_AUTO,
+    1: FAN_LOW,
+}
```
Done in 1feff044ffb5fa9a53c77b06e531587d3216da4d

jesusvila commented:
I've tested this integration and it works perfectly. Aidoo Airzone + Fujitsu ducted air conditioner. The four speeds of the indoor unit are displayed correctly

jesusvila reviewed (approved):
I've tested this integration and it works perfectly. Aidoo Airzone + Fujitsu ducted air conditioner. The four speeds of the indoor unit are displayed correctly

BronRZ reviewed (approved):
I have verified that everything works correctly in my AC with Airzone, I use an LG AC and the operation is perfect, as if it were the official app.

emontnemery commented on homeassistant/components/airzone_cloud/climate.py:
```diff
@@ -290,9 +314,58 @@ def __init__(
+                if speed == 0:
+                    fan_speeds[speed] = FAN_AUTO
+                else:
+                    fan_speeds[speed] = f"{int(round((speed * 100) / max_speed, 0))}%"
```
I suggest styling it like this instead, i.e. with a space before the %-sign:
```suggestion
                    fan_speeds[speed] = f"{int(round((speed * 100) / max_speed, 0))} %"
```

Noltari replied:
I prefer to keep it without a whitespace, since that's how I implemented it in the local integration:
https://github.com/home-assistant/core/blob/0b113b6b714cfa9f3c7421081763dbf8d145b546/homeassistant/components/airzone/climate.py#L169

emontnemery commented on homeassistant/components/airzone_cloud/climate.py:
```diff
@@ -290,9 +314,58 @@ def __init__(
+                else:
+                    fan_speeds[speed] = f"{int(round((speed * 100) / max_speed, 0))}%"
+
+            fan_speeds[1] = FAN_LOW
+            fan_speeds[int(round((max_speed + 1) / 2, 0))] = FAN_MEDIUM
+            fan_speeds[max_speed] = FAN_HIGH
```
Why do we want the list to be:
```
[
        FAN_AUTO,
        FAN_LOW,
        "40%",
        FAN_MEDIUM,
        "80%",
        FAN_HIGH,
    ]
```

Instead of:
```
[
        FAN_AUTO,
        "20 %",
        "40 %",
        "60 %",
        "80 %",
        "100 %",
    ]
```

If there's a good reason, it wouldn't hurt to explain it in a comment.

Noltari replied:
Because only the default speeds are exposed via Homekit/Alexa, and this is the only way of having N speeds and also getting the default ones exposed.

Noltari replied:
Done in 0f4566fd7b73df8542737d0ae8a6c9adc66bc820

emontnemery replied:
> only the default speeds are exposed via Homekit/Alexa

Integrations should not have to solve limitations in voice assistant integrations.
Moreover, the Alexa integration does not support fan speeds for climate entities.

For homekit there is such a limitation though
https://github.com/home-assistant/core/blob/9b1180734a7cfddfe0264591fb00b94a33d557f7/homeassistant/components/homekit/type_thermostats.py#L313-L317

Noltari replied:
> I'd like a second opinion on this design, integrations should not have to solve limitations in voice assistant integrations.

@emontnemery this was already proposed by @bdraco during the same review for the local Airzone integration, see https://github.com/home-assistant/core/pull/92840?w=1#discussion_r1191986188
The Airzone local integration code: https://github.com/home-assistant/core/blob/7c4747bb02e07a126728dbb8fbf5005257a50c9c/homeassistant/components/airzone/climate.py#L158-L177

Noltari replied:
> The Alexa integration does not support fan speeds for climate entities.

You're right, comment fixed in 681967f516b31703c9114f6a36548a20a2215530

emontnemery replied:
> this was already proposed by @bdraco during the same review for the local Airzone integration

That's not a good proposal though, other integrations should not have to implement workarounds because of limitations in `homekit`, please change to a consistent naming both in this PR and in the `airzone` integration.

Noltari replied:
> That's not a good proposal though, other integrations should not have to implement workarounds because of limitations in `homekit`, please change to a consistent naming both in this PR and in the `airzone` integration.

Done in 438c5b6e32483f276b47cb0d5e4747ee5b031d78.
I don't think that doing it for Airzone integration is a good idea since it will break the compatibility for existing users...

emontnemery commented on homeassistant/components/airzone_cloud/climate.py:
```diff
@@ -290,9 +314,58 @@ def __init__(
+            fan_speeds[int(round((max_speed + 1) / 2, 0))] = FAN_MEDIUM
+            fan_speeds[max_speed] = FAN_HIGH
+
+        if 0 in fan_speeds and 0 not in azd_speeds:
+            fan_speeds.pop(0)
```
When does this happen, is it if we pick an item in `FAN_SPEED_MAPS` and the device does not have an automatic speed? Please add a comment.

Noltari replied:
Exactly, it's only for the case that you've mentioned.
I can either add the comment or refactor the code removing 0 from FAN_SPEED_MAPS and adding it only if the device supports the auto speed.

Noltari replied:
Done in 2d63accac765cc94515ca58f2e8ce0af3eef493d
Please let me know if it's more clear now or if we still need a comment for this.

jbouwh replied:
`max_speeds` can be `2` or `3` I guess. But `int(round((max_speed + 1) / 2, 0))` is always `2` if `max_speed` is `2` or `3`. May be the code can be made much easier to read?

emontnemery reviewed (changes requested):
Looks good, just a couple of small issues / unclarities

Noltari committed:
Merge remote-tracking branch 'upstream/dev' into airzone-cloud-climate-speeds

Noltari committed:
airzone_cloud: climate: refactor FAN_AUTO speed

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
airzone_cloud: climate: document standard speeds replacement

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
airzone_cloud: climate: fix Homekit comment

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

Noltari committed:
airzone_cloud: climate: consistent fan speed names

Instead of mixing "low", "medium" and "high" with percentages, let's use only
percentages if there are more speeds.

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

emontnemery reviewed (approved):
Thanks, @Noltari 👍
